### ISSUE TITLE
Can't add items to the list returned by ship:dockingports [BUG]

### ISSUE THREAD

**Issue by nuggreat** [KSP-KOS/KOS#3119]
When you get the list `SHIP:DOCKINGPORTS` you can't add items to that list as you get a "specified cast is not valid" error.
Code to reproduce:
```
SET dl TO SHIP:DOCKINGPORTS.
dl:ADD(0).
```
There is also an assoceated entry in the KSP log from the C# side of the mod
```
[LOG 15:29:23.789] System.InvalidCastException: Specified cast is not valid.
  at (wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
  at kOS.Safe.Encapsulation.Suffixes.OneArgsSuffix`1[TParam].Call (System.Object[] args) [0x00000] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Encapsulation.Suffixes.DelegateSuffixResult.Invoke (kOS.Safe.Execution.ICpu cpu) [0x00226] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Compilation.OpcodeCall.StaticExecute (kOS.Safe.Execution.ICpu cpu, System.Boolean direct, System.Object destination, System.Boolean calledFromKOSDelegateCall) [0x002b2] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Compilation.OpcodeCall.Execute (kOS.Safe.Execution.ICpu cpu) [0x0000d] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Execution.CPU.ExecuteInstruction (kOS.Safe.Execution.ProgramContext context, System.Boolean doProfiling) [0x00203] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Execution.CPU.ContinueExecution (System.Boolean doProfiling) [0x00040] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  at kOS.Safe.Execution.CPU.KOSFixedUpdate (System.Double deltaTime) [0x000b0] in <44fd6c0ba1d84aeaa0b357d09ee83648>:0 
  ```
  which combined with the lack of verbosity makes me thing this is a C# problem getting back into kerboScript some what similar to issue #3006
  
  There is a simple work around where one simply creates a new list of there own and populates it with everything from the dockingports list but that should not be necessary.

**Comment by sug44** [comment]
`ship:dockingports` returns a list that can only hold elements of the docking port type. There are a lot of suffixes and bound variables that return lists like that, and I cant think of a reason for this type of list to be used other than to prevent mistakes in C# code accidentally adding elements of other types. In my opinion the best solution here would be to use `Collections.Generic.List<T>` instead, converting it to kOS list when returning it to user code.

On a side note. Every collection has an overload with a generic type, but besides lists only `UniqueSetValue<UserDelegate>` is being used for various event watchers(undocumented feature afaik).

**Comment by darthgently** [comment]
Does it return a reference to an internal read-only list rather than a modifiable copy?  I’m just shotgunning.  I think there may be built in way to deep copy into your own var if that is what you are getting at.  The way I’m seeing it is that the only reasonable way to add a docking port to ship:dockingports would be to add a docking port to the craft.  I’m not seeming how it would make sense otherwise

**Comment by sug44** [comment]
Yeah you are right. I'm mostly concerned about exposing this weird undocumented type of lists to users. I never tried to do anything to these lists other than indexing or iterating over and I don't think a lot of people did, but still it would be much simpler to return normal lists users can do anything with.

I did some research and right now of all 15 suffixes returning the generic type lists only 3 suffixes don't return a copy(Vessel parts, dockingports, decouplers). When these lists get invalidated(after docking or losing a part for example) the references to previous lists don't get reused and instead new lists are created. This means it's already creating a copy, but only after part invalidations and not on each access to the suffix. One tiny breaking change would be the equality operator returning false instead of true on lists returned by these suffixes between part invalidations.

**Comment by darthgently** [comment]
If the invalidation flag is reliable then getting a reference is ok as long as the user always uses the suffix and not a var.  Seems to me that at minimum the docs could annotate which return copies and which don’t.  I’d have to check the docs, maybe they do, but don’t recall they do.  I always just use the suffix call for that kind of internally managed stuff unless I know I’m using the var within the same tick
